File: apmplugin/src/main/java/com/krubo/apmplugin/timecost/TimeCostClassVisitor.java
package com.krubo.apmplugin.timecost;

import org.objectweb.asm.ClassVisitor;
import org.objectweb.asm.FieldVisitor;
import org.objectweb.asm.MethodVisitor;
import org.objectweb.asm.Opcodes;

public class TimeCostClassVisitor extends ClassVisitor {

    public TimeCostClassVisitor(ClassVisitor classVisitor) {
        super(Opcodes.ASM5, classVisitor);
    }

    @Override
    public MethodVisitor visitMethod(int access, String name, String descriptor, String signature, String[] exceptions) {
        MethodVisitor visitor = super.visitMethod(access, name, descriptor, signature, exceptions);
        return new TimeCostMethodVisitor(visitor, access, name, descriptor);
    }

    @Override
    public FieldVisitor visitField(int access, String name, String descriptor, String signature, Object value) {
        return super.visitField(access, name, descriptor, signature, value);
    }
}

File: apmplugin/src/main/java/com/krubo/apmplugin/timecost/TimeCostClassVisitor.kt
package com.krubo.apmplugin.timecost

import com.krubo.apmplugin.KApmPlugin
import com.krubo.apmplugin.base.BaseClassVisitor
import org.objectweb.asm.ClassVisitor
import org.objectweb.asm.FieldVisitor
import org.objectweb.asm.MethodVisitor

class TimeCostClassVisitor(classVisitor: ClassVisitor) : BaseClassVisitor(classVisitor) {

    override fun checkMethod(className: String, methodName: String?): Boolean {
        println("checkMethod  $className")
        if (KApmPlugin.timeCost.blackMethodClassList.contains("$className/$methodName")){
            println("checkMethod  $className  false  1")
            return false
        }
        if(KApmPlugin.hasInPkgList(className)){
            println("checkMethod  $className  true  1")
            return true
        }
        if (KApmPlugin.timeCost.methodClassList.contains("$className/$methodName")){
            println("checkMethod  $className  true  2")
            return true
        }
        return false
    }

    override fun visitMethod(
        className: String,
        methodName: String?,
        visitor: MethodVisitor?,
        access: Int,
        descriptor: String?,
        signature: String?,
        exceptions: Array<out String>?
    ): MethodVisitor {
        return TimeCostMethodVisitor(className, methodName, visitor, access, descriptor)
    }

    override fun visitField(
        className: String,
        fieldName: String?,
        visitor: FieldVisitor?,
        access: Int,
        descriptor: String?,
        signature: String?,
        value: Any?
    ): FieldVisitor? {
        return null
    }
}